File: app/src/main/java/com/futhark/android/seevoice/controller/adapter/ExerciseListAdapter.java
package com.futhark.android.seevoice.controller.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.futhark.android.seevoice.R;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * exercise list adapter
 * Created by liuhr on 06/04/2017.
 */

public class ExerciseListAdapter<T extends ExerciseListAdapter.ExerciseItemInterface> extends ArrayAdapter<T> {
    private LayoutInflater inflater;
    public ExerciseListAdapter(Context context) {
        super(context, 0);
        this.inflater = LayoutInflater.from(context);
    }

    @SuppressWarnings("unchecked")
    @Override
    @NonNull
    public View getView(int position, View convertView,@NonNull ViewGroup parent) {
        View v = convertView;
        ExerciseItemHolder holder;
        if(v == null){
            v = inflater.inflate(R.layout.item_of_exercise_list, parent, false);
            holder = new ExerciseItemHolder(v);
            v.setTag(holder);
        }else{
            holder = (ExerciseItemHolder)v.getTag();
        }
        holder.setDataToDisplay(getItem(position));
        return v;
    }

    class ExerciseItemHolder{
        @BindView(R.id.item_of_exercise_list_title)
        TextView titleText;
        ExerciseItemHolder(View itemView){
            ButterKnife.bind(this, itemView);
        }

        void setDataToDisplay(T data){
            titleText.setText(data.getTitle());
        }
    }

    public interface ExerciseItemInterface{
        String getTitle();
    }
}

File: app/src/main/java/com/futhark/android/seevoice/controller/adapter/ExerciseListAdapter.kt
package com.futhark.android.seevoice.controller.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView

import com.futhark.android.seevoice.R

/**
 * exercise list adapter
 * Created by liuhr on 06/04/2017.
 */

class ExerciseListAdapter<T : ExerciseListAdapter.ExerciseItemInterface>(context: Context) : ArrayAdapter<T>(context, 0) {
    private val inflater: LayoutInflater

    init {
        this.inflater = LayoutInflater.from(context)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var v = convertView
        val holder: ExerciseItemHolder
        if (v == null) {
            v = inflater.inflate(R.layout.item_of_exercise_list, parent, false)
            holder = ExerciseItemHolder(v)
            v!!.tag = holder
        } else {
            holder = v.tag as ExerciseListAdapter<T>.ExerciseItemHolder
        }
        holder.setDataToDisplay(getItem(position))
        return v
    }

    internal inner class ExerciseItemHolder(itemView: View) {
        var titleText: TextView? = null

        init {
            titleText = itemView.findViewById(R.id.item_of_exercise_list_title)
        }

        fun setDataToDisplay(data: T?) {
            titleText!!.text = data!!.title
        }
    }

    interface ExerciseItemInterface {
        var title: String?
    }
}
